components/brave_sync/client/bookmark_change_processor.cc

```diff
@@ -447,34 +527,265 @@ void BookmarkChangeProcessor::ApplyChangesFromSyncModel(
         if (bookmark_record.isFolder) {
-          node = bookmark_model_->AddFolder(
-                          parent_node,
-                          GetIndex(parent_node, bookmark_record),
-                          base::UTF8ToUTF16(bookmark_record.site.title));
+          node = AddFolder(parent_node,
+                           GetIndex(parent_node, bookmark_record),
+                           base::UTF8ToUTF16(bookmark_record.site.title));
+          folder_was_created = true;
         } else {
-          node = bookmark_model_->AddURL(parent_node,
-                          GetIndex(parent_node, bookmark_record),
-                          base::UTF8ToUTF16(bookmark_record.site.title),
-                          GURL(bookmark_record.site.location));
+          node = AddURL(parent_node,
+                        GetIndex(parent_node, bookmark_record),
+                        base::UTF8ToUTF16(bookmark_record.site.title),
+                        GURL(bookmark_record.site.location));
         }
         if (bookmark_bar_was_empty)
           profile_->GetPrefs()->SetBoolean(bookmarks::prefs::kShowBookmarkBar,
                                           true);
       }
-      UpdateNode(bookmark_model_, node, sync_record.get());
+      UpdateNode(bookmark_model_, node, sync_record.get(), GetPendingNodeRoot());
+
+#ifndef NDEBUG
+      if (parent_node) {
+        ValidateFolderOrders(parent_node);
+      }
+#endif
+
+      if (folder_was_created) {
+        CompletePendingNodesMove(node, sync_record->objectId);
+      }
     }
   }
   bookmark_model_->EndExtensiveChanges();
 }
 
+void BookmarkChangeProcessor::CompletePendingNodesMove(
+    const bookmarks::BookmarkNode* created_folder_node,
+    const std::string& created_folder_object_id) {
+  DCHECK(GetPendingNodeRoot());
+
+  // node, node_order
+  using move_info = std::tuple<bookmarks::BookmarkNode*, const std::string>;
+  std::vector<move_info> move_infos;
+
+  bookmarks::BookmarkNode* pending_node_root = GetPendingNodeRoot();
+  for (int i = 0; i < pending_node_root->child_count(); ++i) {
+    bookmarks::BookmarkNode* node = pending_node_root->GetChild(i);
+
+    std::string parent_object_id;
+    node->GetMetaInfo("parent_object_id", &parent_object_id);
+    if (parent_object_id.empty()) {
+      // The node has been attached to folder which is still in Pending nodes
+      continue;
+    }
+
+    if (created_folder_object_id != parent_object_id) {
+      // Node is still pending, because waits for another parent
+      continue;
+    }
+
+    std::string order;
+    node->GetMetaInfo("order", &order);
+
+    DCHECK(!order.empty());
+    move_infos.push_back(std::make_tuple(node, order));
+  }
+
+  for (auto& move_info : move_infos) {
+    auto* node = std::get<0>(move_info);
+    const auto& order = std::get<1>(move_info);
+    int64_t index = GetIndexByOrder(created_folder_node, order);
+
+    // Use manual move to avoid model observer invocation
```

Comment:
ditto

Reply:
Also the same reply as for AddUrl/AddFolder comment.